# Data Analysis
# Practical 9

# Usual Data loading and processing steps
### Import the libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Read the dataset

In [2]:
kerala_floods = pd.read_csv('kerala_floods.csv')
kerala_floods.head(10)

SUBDIVISION  YEAR   JAN   FEB   MAR    APR    MAY     JUN     JUL    AUG  \
0      KERALA  1901  28.7  44.7  51.6  160.0  174.7   824.6   743.0  357.5   
1      KERALA  1902   6.7   2.6  57.3   83.9  134.5   390.9  1205.0  315.8   
2      KERALA  1903   3.2  18.6   3.1   83.6  249.7   558.6  1022.5  420.2   
3      KERALA  1904  23.7   3.0  32.2   71.5  235.7  1098.2   725.5  351.8   
4      KERALA  1905   1.2  22.3   9.4  105.9  263.3   850.2   520.5  293.6   
5      KERALA  1906  26.7   7.4   9.9   59.4  160.8   414.9   954.2  442.8   
6      KERALA  1907  18.8   4.8  55.7  170.8  101.4   770.9   760.4  981.5   
7      KERALA  1908   8.0  20.8  38.2  102.9  142.6   592.6   902.2  352.9   
8      KERALA  1909  54.1  11.8  61.3   93.8  473.2   704.7   782.3  258.0   
9      KERALA  1910   2.7  25.7  23.3  124.5  148.8   680.0   484.1  473.8   

     SEP    OCT    NOV    DEC   ANNUAL RAINFALL FLOODS  
0  197.7  266.9  350.8   48.4            3248.6    YES  
1  491.6  358.4  158.3  121.5            3326.6    YES  
2  341.8  354.1  157.0   59.0            3271.2    YES  
3  222.7  328.1   33.9    3.3            3129.7    YES  
4  217.2  383.5   74.4    0.2            2741.6     NO  
5  131.2  251.7  163.1   86.0            2708.0     NO  
6  225.0  309.7  219.1   52.8            3671.1    YES  
7  175.9  253.3   47.9   11.0            2648.3     NO  
8  195.4  212.1  171.1   32.3            3050.2    YES  
9  248.6  356.6  280.4    0.1            2848.6     NO

## Size of the dataset

In [3]:
kerala_floods.shape

(118, 16)

## Missing values

In [4]:
kerala_floods.isna().sum()

SUBDIVISION         0
YEAR                0
JAN                 0
FEB                 0
MAR                 0
APR                 0
MAY                 0
JUN                 0
JUL                 0
AUG                 0
SEP                 0
OCT                 0
NOV                 0
DEC                 0
 ANNUAL RAINFALL    0
FLOODS              0
dtype: int64

# 1 - New column JUNE_MORE_THAN_500 

In [5]:
kerala_floods['JUNE_MORE_THAN_500'] = (kerala_floods['JUN'] > 500).astype(int)

In [6]:
kerala_floods['JUNE_MORE_THAN_500'].value_counts()

JUNE_MORE_THAN_500
1    93
0    25
Name: count, dtype: int64

# 2 - Updates to the column 'FLOODS'

In [7]:
kerala_floods['FLOODS'].value_counts()

FLOODS
YES    60
NO     58
Name: count, dtype: int64

In [8]:
kerala_floods['FLOODS'] = (kerala_floods['FLOODS'] == 'YES').astype(int)

In [9]:
kerala_floods['FLOODS'].value_counts()

FLOODS
1    60
0    58
Name: count, dtype: int64

# 3 - Select needed columns

In [10]:
kerala_floods[['JUNE_MORE_THAN_500', 'FLOODS']]

JUNE_MORE_THAN_500  FLOODS
0                     1       1
1                     0       1
2                     1       1
3                     1       1
4                     1       0
..                  ...     ...
113                   0       1
114                   1       0
115                   1       0
116                   0       0
117                   1       1

[118 rows x 2 columns]

# 3 - Crosstab between these two columns: FLOODS, JUNE_MORE_THA_500

In [11]:
crosstab = pd.crosstab(kerala_floods['JUNE_MORE_THAN_500'], kerala_floods['FLOODS'])

In [12]:
crosstab

FLOODS               0   1
JUNE_MORE_THAN_500        
0                   19   6
1                   39  54

# 4 -  P(JUNE_MORE_THAN_500) = ? 

In [13]:
june_more_than_500 = kerala_floods['JUNE_MORE_THAN_500'].value_counts()[1] 
june_more_than_500

np.int64(93)

In [14]:
all_june = kerala_floods['JUNE_MORE_THAN_500'].count()
all_june

np.int64(118)

In [15]:
p_june_more_than_500 = june_more_than_500/all_june
p_june_more_than_500

np.float64(0.788135593220339)

### P(JUNE & RAIN>500) = 0.7881

# 5 - P(FLOOD)

In [16]:
flood = kerala_floods['FLOODS'].value_counts()[1]

In [17]:
flood

np.int64(60)

In [18]:
all_flood = kerala_floods['FLOODS'].count()
all_flood

np.int64(118)

In [19]:
P_flood = flood / all_flood
P_flood

np.float64(0.5084745762711864)

### P(FLOOD) = 0.5085

# 50% probability of flood, that's actually crazy 

# 6 - P(FLOOD & RAIN > 500 IN JUNE)

In [53]:
cnt_fl_r_500_june = ((kerala_floods['JUNE_MORE_THAN_500'] == 1) & (kerala_floods['FLOODS'] == 1)).sum()
cnt_fl_r_500_june

np.int64(54)

In [55]:
N = len(kerala_floods)
N

118

In [56]:
P_fl_r_500_june = cnt_fl_r_500_june / N
P_fl_r_500_june

np.float64(0.4576271186440678)

### P(FLOOD & RAIN > 500 IN JUNE) = 0.4

# 7 - P(FLOOD | RAIN > 500 in JUNE)

In [57]:
P_fl_given_r_500_june = P_fl_r_500_june / p_june_more_than_500
P_fl_given_r_500_june

np.float64(0.5806451612903226)

### P(FLOOD|RAIN>500 IN JUNE) = 58.0645

### So in this region, we are expecting to see a flood in the month of june if it rains 500 mm for 50%, so if it is heavy raining at morning and you see 50mm in your bucket, I think it is good idea to pack your things. 

# 8 - P(RAIN < 500 IN JUNE)

In [22]:
P_r_500less_june = kerala_floods['JUNE_MORE_THAN_500'].value_counts()[0] / all_june
P_r_500less_june

np.float64(0.211864406779661)

# P(RAIN < 500 IN JUNE) = 0.2119

### That's some rainy place

# 9 - P(FLOOD | RAIN < 500 IN JUNE )

### Intuitively this should be way less than what we have before. 

In [23]:
P_fl_given_r_500less_june = P_flood / P_r_500less_june
P_fl_given_r_500less_june

np.float64(2.4)

### I must have confused few things. 
### P(A|B) = P(A AND B) / P(B)

In [24]:
kerala_floods['JUNE_MORE_THAN_500'].value_counts()[0]

np.int64(25)

In [25]:
P_fl_r_500less_june = (flood / all_flood) * (kerala_floods['JUNE_MORE_THAN_500'].value_counts()[0] / all_flood) 

### Now, this is the multipilcation of flood and rain < 500 in june and I have to divide this value by P(rain < 500 in June)
### I could also write it this way

In [26]:
P_fl_r_500less_june = P_flood * P_r_500less_june
P_fl_r_500less_june

np.float64(0.1077276644642344)

In [27]:
P_fl_given_r_500less_june = P_fl_r_500less_june / P_r_500less_june
P_fl_given_r_500less_june

np.float64(0.5084745762711864)

### OK, I got smth that looks more realistic, but still not satisfied. Should I multiply probability of two events or add them?

### I believe, I realized my mistakes

In [35]:
cnt_500less_june = kerala_floods['JUNE_MORE_THAN_500'].value_counts()[0]
cnt_500less_june

np.int64(25)

In [37]:
cnt_flood = kerala_floods['FLOODS'].value_counts()[1]
cnt_flood

np.int64(60)

### So, first I need to find P(FLOODS & RAIN < 500 IN JUNE)

In [43]:
cnt_r_500less_june_fl = ((kerala_floods['JUNE_MORE_THAN_500'] < 1) & (kerala_floods['FLOODS'] > 0)).sum() 
cnt_r_500less_june_fl

np.int64(6)

In [50]:
P_fl_r_500less_june = cnt_r_500less_june_fl / N
P_fl_r_500less_june

np.float64(0.05084745762711865)

### We have 5 % prob that there is less than 500 mm of rain in june and having flood the same year. No we divide that by P(RAIN < 500 IN JUNE) 

In [51]:
P_fl_given_r_500less_june = P_fl_r_500less_june / P_r_500less_june
P_fl_given_r_500less_june

np.float64(0.24000000000000002)

### YES, I GOT THIS! We have 24% prob that there's gonna be flood given there is less than 500 mm of rain. People of that region is not getting their peace. 

## Last task is to find probability of having rainfall more than 500 mm in june, given that it has flooded this year.

# 10 - Bayes' Theorem and Total Probability

### Now, that I figured out the correct conditional probability, we can easily find the "reverse" condition. Just stick to this formula
### P(R<500 IN JUNE| FLOOD) = P(FLOOD|R<500 IN JUNE) * P(R<500 IN JUNE) / P(FLOOD)

In [52]:
P_r_500less_june_given_fl = P_fl_given_r_500less_june * P_r_500less_june / P_flood
P_r_500less_june_given_fl

np.float64(0.1)

### Total probability

$P(R<500) = P(R<500|F) * P(F) + P(R<500|F^c) * P(F^c) $

$P(R<500|F^c) = P(R<500  \cap F^c) / P(F^c) $

In [58]:
cnt_no_flood = (kerala_floods['FLOODS'] == 0).sum()
cnt_no_flood

np.int64(58)

In [ ]:
P_no_flood = cnt_no_flood / N
P_no_flood 

np.float64(0.4915254237288136)

In [60]:
cnt_r_500less_june_no_fl = ((kerala_floods['JUNE_MORE_THAN_500'] == 0) & (kerala_floods['FLOODS'] == 0)).sum()
cnt_r_500less_june_no_fl

np.int64(19)

In [61]:
P_r_500less_june_no_fl = cnt_r_500less_june_no_fl / N
P_r_500less_june_no_fl

np.float64(0.16101694915254236)

In [62]:
P_r_500less_june_given_no_fl = P_r_500less_june_no_fl / P_no_flood
P_r_500less_june_given_no_fl

np.float64(0.32758620689655166)

$P(R<500) = P(R<500|F) * P(F) + P(R<500|F^c) * P(F^c) $

In [63]:
P_r_500less = P_r_500less_june_given_fl * P_flood + P_r_500less_june_given_no_fl * P_no_flood
P_r_500less

np.float64(0.211864406779661)

# 11 - Bayes' Thoerem

### Using Bayes' Theorem, we could easily find our conditional probability. This is very useful especially in the context of journalisms. Because, they usually twist the plots of their talk to lure people into their post. But we could easily use bayes theorem to prove their wrong. Maybe, just prove the reverse the conditional probability. 

### Thank you! 